File: apps/ios/InfraAuth/Todo/TodoViewModel.swift
import Foundation

/// Owns the current user's todo list plus its create/toggle/delete mutations —
/// the Swift counterpart of web's `useTodos` hook. Mutations update local state
/// from the server's returned DTO (no full re-fetch), keeping the list
/// authoritative without a flash. All state mutates on the main actor.
@MainActor
final class TodoViewModel: ObservableObject {
    /// nil until the first load resolves; drives the loading skeleton.
    @Published private(set) var todos: [TodoDTO]?
    @Published private(set) var error: String?
    /// True while a create is in flight (disables the add form).
    @Published private(set) var creating = false
    /// Ids with a toggle/delete in flight (disables that row).
    @Published private(set) var pendingIds: Set<String> = []

    private let client: TodoClient

    init(client: TodoClient) {
        self.client = client
    }

    var loading: Bool { error == nil && todos == nil }

    /// Load the list once. Idempotent-ish: re-runs only refetch on demand.
    func load() async {
        do {
            todos = try await client.list()
        } catch {
            self.error = "无法加载待办，请稍后重试。"
        }
    }

    func create(title: String) async {
        let normalized = TodoValidation.normalize(title)
        guard !normalized.isEmpty else { return }
        error = nil
        creating = true
        defer { creating = false }
        do {
            let created = try await client.create(title: normalized)
            // List is newest-first; the new item leads.
            todos = [created] + (todos ?? [])
        } catch {
            self.error = "创建失败，请重试。"
        }
    }

    func toggle(_ todo: TodoDTO) async {
        await withPending(todo.id) {
            self.error = nil
            do {
                let updated = try await self.client.toggle(id: todo.id, completed: !todo.completed)
                self.todos = self.todos?.map { $0.id == updated.id ? updated : $0 }
            } catch {
                self.error = "更新失败，请重试。"
            }
        }
    }

    func remove(id: String) async {
        await withPending(id) {
            self.error = nil
            do {
                try await self.client.remove(id: id)
                self.todos = self.todos?.filter { $0.id != id }
            } catch {
                self.error = "删除失败，请重试。"
            }
        }
    }

    private func withPending(_ id: String, _ body: () async -> Void) async {
        pendingIds.insert(id)
        defer { pendingIds.remove(id) }
        await body()
    }
}

File: apps/android/app/src/main/kotlin/dev/w3ctech/infralab/ui/todos/TodoViewModel.kt
package dev.w3ctech.infralab.ui.todos

import dev.w3ctech.infralab.data.TodoClient
import dev.w3ctech.infralab.data.contracts.TodoDTO
import dev.w3ctech.infralab.data.contracts.TodoInput
import dev.w3ctech.infralab.di.ServiceLocator
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Owns the current user's todo list plus its create/toggle/delete mutations — the Kotlin
 * counterpart of web's `useTodos` hook. Mutations update local state from the server's returned
 * DTO (no full re-fetch), keeping the list authoritative without a flash.
 */
data class TodoUiState(
    /** null until the first load resolves; drives the loading skeleton. */
    val todos: List<TodoDTO>? = null,
    val error: String? = null,
    /** True while a create is in flight (disables the add form). */
    val creating: Boolean = false,
    /** Ids with a toggle/delete in flight (disables that row). */
    val pendingIds: Set<String> = emptySet(),
) {
    val loading: Boolean get() = error == null && todos == null
}

class TodoViewModel(private val client: TodoClient) : ViewModel() {

    private val _state = MutableStateFlow(TodoUiState())
    val state: StateFlow<TodoUiState> = _state.asStateFlow()

    fun load() {
        viewModelScope.launch {
            runCatching { client.list() }
                .onSuccess { todos -> _state.update { it.copy(todos = todos, error = null) } }
                .onFailure { _state.update { it.copy(error = "无法加载待办，请稍后重试。") } }
        }
    }

    fun create(rawTitle: String) {
        val title = TodoInput.normalizeTitle(rawTitle)
        if (title.isEmpty()) return
        viewModelScope.launch {
            _state.update { it.copy(creating = true, error = null) }
            runCatching { client.create(title) }
                .onSuccess { created ->
                    // List is newest-first; the new item leads.
                    _state.update { it.copy(creating = false, todos = listOf(created) + (it.todos ?: emptyList())) }
                }
                .onFailure { _state.update { it.copy(creating = false, error = "创建失败，请重试。") } }
        }
    }

    fun toggle(todo: TodoDTO) {
        withPending(todo.id) {
            runCatching { client.toggle(todo.id, !todo.completed) }
                .onSuccess { updated ->
                    _state.update { s -> s.copy(todos = s.todos?.map { if (it.id == updated.id) updated else it }) }
                }
                .onFailure { _state.update { it.copy(error = "更新失败，请重试。") } }
        }
    }

    fun remove(id: String) {
        withPending(id) {
            runCatching { client.remove(id) }
                .onSuccess { _state.update { s -> s.copy(todos = s.todos?.filter { it.id != id }) } }
                .onFailure { _state.update { it.copy(error = "删除失败，请重试。") } }
        }
    }

    private fun withPending(id: String, block: suspend () -> Unit) {
        viewModelScope.launch {
            _state.update { it.copy(pendingIds = it.pendingIds + id, error = null) }
            try {
                block()
            } finally {
                _state.update { it.copy(pendingIds = it.pendingIds - id) }
            }
        }
    }

    companion object {
        val Factory: ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T =
                TodoViewModel(ServiceLocator.todoClient()) as T
        }
    }
}
